src/frontend/planner_test/tests/testdata/output/join.yaml

```diff
@@ -506,27 +506,29 @@
           ├─LogicalScan { table: a, columns: [a.x, a._row_id] }
           └─LogicalScan { table: b, columns: [b.x, b._row_id] }
   batch_plan: |-
-    BatchProject { exprs: [$expr1, $expr2] }
-    └─BatchExchange { order: [$expr3 ASC], dist: Single }
-      └─BatchSort { order: [$expr3 ASC] }
-        └─BatchProject { exprs: [(2:Int32 * Coalesce(a.x, b.x)) as $expr1, (Coalesce(a.x, b.x) + Coalesce(a.x, b.x)) as $expr2, (Coalesce(a.x, b.x) + Coalesce(a.x, b.x)) as $expr3] }
-          └─BatchFilter { predicate: ((2:Int32 * Coalesce(a.x, b.x)) < 10:Int32) }
-            └─BatchHashJoin { type: FullOuter, predicate: a.x = b.x, output: all }
-              ├─BatchExchange { order: [], dist: HashShard(a.x) }
-              │ └─BatchScan { table: a, columns: [a.x], distribution: SomeShard }
-              └─BatchExchange { order: [], dist: HashShard(b.x) }
-                └─BatchScan { table: b, columns: [b.x], distribution: SomeShard }
+    BatchProject { exprs: [$expr3, $expr2] }
+    └─BatchExchange { order: [$expr2 ASC], dist: Single }
+      └─BatchProject { exprs: [(2:Int32 * $expr1) as $expr3, $expr2, $expr2] }
+        └─BatchSort { order: [$expr2 ASC] }
+          └─BatchProject { exprs: [a.x, b.x, Coalesce(a.x, b.x) as $expr1, (Coalesce(a.x, b.x) + Coalesce(a.x, b.x)) as $expr2] }
```

Comment:
Seems `Coalesce(a.x, b.x)` is not extracted here?

Reply:
Good Catch! We should apply this rule in top down order.